api/service/legacysync/epoch_syncing.go

```diff
@@ -0,0 +1,426 @@
+package legacysync
+
+import (
+	"sync"
+	"time"
+
+	"github.com/Workiva/go-datastructures/queue"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/common/math"
+	"github.com/harmony-one/harmony/api/service/legacysync/downloader"
+	"github.com/harmony-one/harmony/consensus"
+	"github.com/harmony-one/harmony/consensus/engine"
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/core/types"
+	"github.com/harmony-one/harmony/internal/chain"
+	"github.com/harmony-one/harmony/internal/utils"
+	"github.com/harmony-one/harmony/node/worker"
+	"github.com/harmony-one/harmony/p2p"
+	"github.com/harmony-one/harmony/shard"
+	"github.com/pkg/errors"
+)
+
+type EpochSync struct {
+	blockChain         *core.BlockChain
+	selfip             string
+	selfport           string
+	selfPeerHash       [20]byte // hash of ip and address combination
+	commonBlocks       map[int]*types.Block
+	lastMileBlocks     []*types.Block // last mile blocks to catch up with the consensus
+	syncConfig         *SyncConfig
+	isExplorer         bool
+	stateSyncTaskQueue *queue.Queue
+	syncMux            sync.Mutex
+	lastMileMux        sync.Mutex
+
+	syncStatus syncStatus
+}
+
+// GetSyncStatus get the last sync status for other modules (E.g. RPC, explorer).
+// If the last sync result is not expired, return the sync result immediately.
+// If the last result is expired, ask the remote DNS nodes for latest height and return the result.
+func (ss *EpochSync) GetSyncStatus() SyncCheckResult {
+	return ss.syncStatus.Get(func() SyncCheckResult {
+		return ss.isInSync(false)
+	})
+}
+
+// isInSync query the remote DNS node for the latest height to check what is the current
+// sync status
+func (ss *EpochSync) isInSync(doubleCheck bool) SyncCheckResult {
```

Comment:
Whether epoch blocks are in sync should depends on checking the latest epoch block right? Seems this part of the code is just a copy and paste of the logic that's checking the latest block of the chain?

Reply:
Right, now it depends on last epoch.

Reply:
I think for checking whether the epoch blocks are latest in sync with the beacon chain, the check logic should be:

1. Get the latest epoch block I have: myEpoch.
2. Get maxPeerHeight: otherHeight
3. using shard schedule to calculate what epoch is the "otherHeight" belong and get: otherEpoch.
4. Compare otherEpoch vs myEpoch to decide whether I am in sync or not.

Reply:
OnTop means i have the highest block that possible. In epoch context it means that i have the highest epoch. I can calculate last block for this epoch, and if it doesn't exists yet means i'm idling. Otherwise i'm synchronizing. I do the same above, but in another words, without converting into epoch, but in heights.

Reply:
i've realized that "in sync" means already synchronized but not "in synchronizing process", exactly opposite meaning. Fixed.